pkg/controller/storagecluster/storagecluster.go

```diff
@@ -1176,22 +1186,91 @@ func (c *Controller) getCurrentMaxStorageNodesPerZone(
 	return 0, fmt.Errorf("storage disabled")
 }
 
+func getDefaultStorageNodesDisaggregatedMode(
+	nodeList *v1.NodeList,
+	cluster *corev1.StorageCluster,
+	recorder record.EventRecorder,
+) (uint64, bool, error) {
+	// Check if ENABLE_ASG_STORAGE_PARTITIONING is set to false. If not, we can look at the 'portworx.io/node-type' label
+	for _, envVar := range cluster.Spec.Env {
+		if envVar.Name == StoragePartitioningEnvKey {
+			if envVar.Value == "false" {
+				return 0, false, nil
+			}
+			break
+		}
+	}
+
+	// This will return a zone map of storage nodes in each zones
+	nodeTypeZoneMap := getZoneMap(nodeList, NodeTypeKey, StorageNodeValue)
+	if nodeTypeZoneMap == nil {
+		return 0, false, fmt.Errorf("count not find zone information")
+	}
+	totalNodes := uint64(0)
+	// We'll find the zone with least number of labels
+	minValue := uint64(math.MaxUint64)
+	prevKey := ""
+	prevValue := uint64(0)
+	for key, value := range nodeTypeZoneMap {
+		totalNodes += value
+		if value < minValue {
+			minValue = value
+		}
+		if prevValue != uint64(0) && prevValue != value {
+			k8s.InfoEvent(
+				recorder, cluster, util.UnevenStorageNodesReason,
+				fmt.Sprintf("Uneven number of storage nodes labelled across zones."+
+					" %v has %v, %v has %v", prevKey, prevValue, key, value),
+			)
+		}
+		prevKey = key
+		prevValue = value
+	}
+	if totalNodes == 0 {
+		// no node is labelled with portworx.io/node-type
+		nodeTypeZoneMap = getZoneMap(nodeList, NodeTypeKey, StoragelessNodeValue)
+		for _, value := range nodeTypeZoneMap {
+			totalNodes += value
+		}
+		if totalNodes > 0 {
+			k8s.InfoEvent(
+				recorder, cluster, util.AllStoragelessNodesReason,
+				fmt.Sprintf("%v nodes marked as storageless, none marked as storage nodes", totalNodes),
+			)
+			return 0, true, fmt.Errorf("storageless nodes found. None marked as storage node")
+		}
+		return minValue, false, nil
```

Comment:
minValue can be used incorrectly by callers, why not just return 0? otherwise we can call `getZoneMap(nodeList, "", "")` again then calculate the number of storage nodes? that case we don't even need the bool return value in this function and still have a single function `getDefaultMaxStorageNodesPerZone`

Reply:
>> minValue can be used incorrectly by callers, why not just return 0?
I kept minValue since totalNodes is 0, minValue will be zero as well. I made it explicit to be clear. 

For you suggestion on merging with the caller: I feel we should keep the two decisions separate as modules. That will help us reuse this function in an integration test later.